detection_rules/rule.py

```diff
@@ -438,13 +439,30 @@ def autobumped_version(self) -> Optional[int]:
 
         return version + 1 if self.is_dirty else version
 
-    @staticmethod
-    def _post_dict_transform(obj: dict) -> dict:
+    def _post_dict_transform(self, obj: dict) -> dict:
         """Transform the converted API in place before sending to Kibana."""
 
         # cleanup the whitespace in the rule
         obj = nested_normalize(obj)
 
+        if not isinstance(self, DeprecatedRuleContents):
+            current_version = Version(load_current_package_version())
+            restricted_fields = self.data.get_restricted_fields
+
+            for field_name, stack_values in restricted_fields.items():
```

Comment:
I think we can bypass iterating through each restricted field since in this case we _know_ what we want to act on. So we can simply act on it only if it is defined (and within the bounds of stack versions expected).

We should also break out this logic to a dedicated method under `TOMLRuleContents` since it is a bit complex and not to muddle with other build time logic.

defined as a global constant pattern, since we will only build it as long as we need to (not necessarily for the full duration of validity as a field):

```python
BUILD_FIELD_REQUIRED_FIELDS = (Version('8.3'), None)
```

```suggestion
    def _build_required_fields(self, obj: dict) -> dict:
         """Derive required fields from query AST."""
         min_stack, max_stack = BUILD_FIELD_REQUIRED_FIELDS
         max_stack = max_stack or current_version

         if Version(min_stack) <= current_version >= Version(max_stack):
             obj.setdefault(field_name, obj.get(field_name, []))
```

I have been thinking about it and I do think we need to do a validation of the generated field against the schema just to be safe. We should be able to pull out the field from the `data` schema, or just generate a marshmallow field directly off the dataclass and use the predefined validation in place (in this case is `Optional[List[str]]`

Reply:
Is a global variable of `BUILD_FIELD_REQUIRED_FIELDS` necessary since we already define the min for each field in `BaseRuleData` schema?  If our global variable specifying `8.3` for each new field in a rule, it may become redundant as it has already specified a min_stack_version in `BaseRuleData` schema and that is pulled in `get_restricted_fields`.

Reply:
Didn't need the `BUILD_FIELD_REQUIRED_FIELDS` since it is redundant to `get_restricted_fields` which pulls from the schema. Moving forward, we won't have to add additional global variables.